pkg/apis/v1/labels.go

```diff
@@ -104,38 +99,72 @@ var (
 	}
 )
 
-// IsRestrictedLabel returns an error if the label is restricted.
+// IsRestrictedLabel is used for runtime validation of requirements.
+// Returns an error if the label is restricted. E.g. using .karpenter.sh suffix.
 func IsRestrictedLabel(key string) error {
 	if WellKnownLabels.Has(key) {
 		return nil
 	}
-	if IsRestrictedNodeLabel(key) {
+
+	labelDomain := GetLabelDomain(key)
+	for restrictedLabelDomain := range RestrictedLabelDomains {
+		if labelDomain == restrictedLabelDomain || strings.HasSuffix(labelDomain, "."+restrictedLabelDomain) {
+			return fmt.Errorf("label %s is restricted; specify a well known label: %v, or a custom label that does not use a restricted domain: %v", key, sets.List(WellKnownLabels), sets.List(RestrictedLabelDomains))
+		}
+	}
+
+	if RestrictedLabels.Has(key) {
 		return fmt.Errorf("label %s is restricted; specify a well known label: %v, or a custom label that does not use a restricted domain: %v", key, sets.List(WellKnownLabels), sets.List(RestrictedLabelDomains))
 	}
+
 	return nil
 }
 
-// IsRestrictedNodeLabel returns true if a node label should not be injected by Karpenter.
-// They are either known labels that will be injected by cloud providers,
-// or label domain managed by other software (e.g., kops.k8s.io managed by kOps).
-func IsRestrictedNodeLabel(key string) bool {
+// IsValidLabelToSync returns true if the label key is allowed to be synced to the Node object centrally by Karpenter.
+func IsValidToSyncCentrallyLabel(key string) bool {
+	// TODO(enxebre): consider this to be configurable with runtime flag.
+	notValidToSyncLabel := WellKnownLabels
+
+	return !notValidToSyncLabel.Has(key)
+}
+
+// IsKubeletLabel returns true if the label key is one that kubelets are allowed to set on their own Node object.
+// This function is similar the one used by the node restriction admission https://github.com/kubernetes/kubernetes/blob/e319c541f144e9bee6160f1dd8671638a9029f4c/staging/src/k8s.io/kubelet/pkg/apis/well_known_labels.go#L67
+// but karpenter also restricts the known labels to be passed to kubelet. Only the kubeletLabelNamespaces are allowed.
+func IsKubeletLabel(key string) bool {
 	if WellKnownLabels.Has(key) {
+		return false
+	}
+
+	if !isKubernetesLabel(key) {
 		return true
 	}
-	labelDomain := GetLabelDomain(key)
-	for exceptionLabelDomain := range LabelDomainExceptions {
-		if strings.HasSuffix(labelDomain, exceptionLabelDomain) {
-			return false
+
+	namespace := GetLabelDomain(key)
+	for allowedNamespace := range kubeletLabelNamespaces {
+		if namespace == allowedNamespace || strings.HasSuffix(namespace, "."+allowedNamespace) {
+			return true
 		}
 	}
-	for restrictedLabelDomain := range RestrictedLabelDomains {
-		if strings.HasSuffix(labelDomain, restrictedLabelDomain) {
+
+	return false
+}
+
+func isKubernetesLabel(key string) bool {
+	for k8Domain := range K8sLabelDomains {
```

Comment:
This is just a nit pick, but shouldn't `k8Domain` be called `k8sDomain`?  "K8s" is short for "kubernetes" while "k8" would be short for "kubernete".  Plus using "k8s" matches the name used elsewhere in this code.